File: files/CH-202.py
import pandas as pd
import numpy as np

path = "CH-202 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="B", skiprows=1, nrows=17)
test = pd.read_excel(path, usecols="D:F", skiprows=1, nrows=9)

input['Date'] = input['Column 1'].where(pd.to_datetime(input['Column 1'], errors='coerce').notna())
input['Date'] = input['Date'].apply(lambda x: np.nan if pd.to_datetime(x, errors='coerce') and len(str(x)) == 1 else x)
input['Date'] = input['Date'].ffill()
input = input[input['Date'] != input['Column 1']]
input['Type'] = input['Column 1'].apply(lambda x: 'Quantity' if str(x).isdigit() else 'Product')
input['change'] = np.where((input['Type'].shift() == "Product") & (input['Type'] == "Quantity"), 0, 1)
input['row'] = input.groupby('Date')['change'].cumsum()
result = input.pivot(index=['Date', 'row'], columns='Type', values='Column 1').reset_index()
result['Quantity'] = result['Quantity'].astype('float64')
result.columns.name = None
result = result.drop(columns=['row'])
print(result.equals(test)) # True

File: files/200-299/202/CH-202.R
library(tidyverse)
library(readxl)

path = "files/CH-202 Table Transformation.xlsx"
input = read_excel(path, range = "B2:B18")
test  = read_excel(path, range = "D2:F11")

result = input %>%
  mutate(Date = if_else(str_detect(`Column 1`, "^\\d+$") & as.numeric(`Column 1`) > 40000,
                        as_date(as.numeric(`Column 1`), origin = "1899-12-30"),
                        NA_Date_) %>% as.POSIXct()) %>%
  fill(Date) %>%
  filter(nchar(`Column 1`) != 5) %>%
  mutate(col = if_else(str_detect(`Column 1`, "^\\d+$"),"Quantity","Product")) %>%
  mutate(change = ifelse(lag(col) == "Product" & col == "Quantity", 0, 1), .by = Date) %>%
  group_by(Date) %>%
  mutate(row = cumsum(change)) %>%
  ungroup() %>%
  select(-change) %>%
  pivot_wider(names_from = col, values_from = `Column 1`) %>%
  select(-row) %>%
  mutate(Quantity = as.numeric(Quantity))


all.equal(result, test)
#> [1] TRUE